File: 20201208_day_8/rust/src/instruction.rs
// Struct that defines an 'instruction', i.e. a line from the input file
#[derive(Debug)]
pub enum Instruction {
    ACC(i32),
    JMP(i32),
    NOP(i32),
}

impl Instruction {
    // Create an Instruction from an input file line
    pub fn from_string(s: &str) -> Instruction {
        let instruction = s[..4].trim(); // The alpha part
        let value = s[4..].trim().parse().unwrap(); // The numeric part

        // Convert to Instruction based on the alpha part
        match instruction {
            "acc" => Instruction::ACC(value),
            "jmp" => Instruction::JMP(value),
            "nop" => Instruction::NOP(value),
            _ => panic!("Can't parse Instruction from {}", s),
        }
    }

    // Given a mutable reference to a line number in the instruction set `pointer` and a mutable
    // reference to an accumulator `acc`, update `pointer` and `acc` based on the desired
    // Instruction behavior (i.e., Instruction::ACC moves one instruction forward and adds its
    // value to the accumulator).
    pub fn execute(&self, pointer: &mut usize, acc: &mut i32) {
        match self {
            Instruction::ACC(x) => {
                *pointer += 1;
                *acc += x;
            }
            Instruction::JMP(x) => *pointer = (*pointer as i32 + x) as usize,
            Instruction::NOP(_) => *pointer += 1,
        }
    }

    // Given a mutable reference to a line number in the instruction set `pointer`, run the
    // Instruction (JMP or NOP) as if it were the other one (NOP or JMP). Ignore ACC instructions
    pub fn flip_execute(&self, pointer: &mut usize) {
        match self {
            Instruction::ACC(_) => (),
            Instruction::JMP(x) => Instruction::NOP(*x).execute(pointer, &mut 0),
            Instruction::NOP(x) => Instruction::JMP(*x).execute(pointer, &mut 0),
        }
    }
}

File: 20201208_day_8/dart/bin/instruction.dart
import 'dart:math';

enum InstructionType {
  acc,
  jmp,
  nop,
}

class Instruction {
  static final reFromString = RegExp(r'^(nop|jmp|acc)\s*(\+|\-)(\d+)$');

  late int id;
  late InstructionType type;
  late int value;

  Instruction(InstructionType type, int value) {
    id = Random().nextInt(1000);
    this.type = type;
    this.value = value;
  }

  Instruction.fromString(String s) {
    id = Random().nextInt(1000);
    var stringMatch = reFromString.firstMatch(s);
    if (stringMatch == null) {
      throw ArgumentError('Cannot parse $s into Instruction');
    }

    switch (stringMatch.group(1)!) {
      case 'acc':
        type = InstructionType.acc;
        break;
      case 'jmp':
        type = InstructionType.jmp;
        break;
      case 'nop':
        type = InstructionType.nop;
        break;
      default:
        throw ArgumentError('Cannot parse $s into Instruction');
    }

    var sign = 1;
    if (stringMatch.group(2)! == '-') {
      sign = -1;
    }
    value = int.parse(stringMatch.group(3)!) * sign;
  }

  @override
  bool operator ==(Object other) {
    return other is Instruction &&
        id == other.id &&
        type == other.type &&
        value == other.value;
  }

  @override
  int get hashCode {
    return id.hashCode ^ type.hashCode ^ value.hashCode;
  }
}
